packages/affiliates/gas-rebate/VoterGasRebateV2.ts

```diff
@@ -0,0 +1,128 @@
+// This is a script calculates the gas rebate for UMA protocol voters. The script finds all VoteCommitted and
+// VoteRevealed events in a specified time range and calculates the gas used for each event. It then uses UniswapV3 swap
+// events to find the UMA/ETH price at the block in which each event was included and calculates the ETH cost of the gas
+// used in UMA tokens. The script aggregates the gas rebates by voter and saves the results to a file. // It is designed
+// to not require any run time parameters by always running it one month after the desired output month.
+
+import { getAddress } from "@uma/contracts-node";
+const hre = require("hardhat");
+const { ethers } = hre;
+import { BigNumber } from "ethers";
+import bn from "bignumber.js";
+const moment = require("moment");
+import { findBlockNumberAtTimestamp, getWeb3, decodePriceSqrt } from "@uma/common";
+import fs from "fs";
+import path from "path";
+
+// Find the closest UMA/ETH price to the given block number by searching through swap events in the given range.
+function findClosestUmaEthPrice(swapEvents: any, blockNumber: number) {
+  const closest = swapEvents.reduce((prev: any, curr: any) => {
+    return Math.abs(curr.blockNumber - blockNumber) < Math.abs(prev.blockNumber - blockNumber) ? curr : prev;
+  });
+  // The returned price is of the form 1 UMA = x ETH. We want to return the price of 1 ETH in UMA. Also, Uniswap returns
+  // the price in x96 square root encoded. We need to decode by reverse x96 operation by compting (price/(2^96))^2.
+  const basePrice = decodePriceSqrt(closest.args.sqrtPriceX96.toString());
+  const bnPrice = new bn(ethers.utils.parseUnits("1").toString()).div(new bn(basePrice.toString()));
+  return BigNumber.from(bnPrice.toString().toString().substring(0, bnPrice.toString().indexOf(".")));
+}
+
+export async function run(): Promise<void> {
+  console.log("Running UMA2.0 Gas rebate script! This script assumes you are running it for the previous month🍌.");
+
+  // Work out the range to run over. This should be over the totality of the previous month.
+  const currentDate = new Date();
+  const prevMonthStart = new Date(currentDate);
+  prevMonthStart.setMonth(prevMonthStart.getMonth() - 1);
+  prevMonthStart.setDate(1);
+  prevMonthStart.setHours(0, 0, 0, 0);
+
+  const prevMonthEnd = new Date(prevMonthStart.getFullYear(), prevMonthStart.getMonth() + 1, 0);
+  prevMonthEnd.setHours(23, 59, 59);
```

Comment:
i think this date math will be timezone dependent depending on where script is run. probably not a big deal, but if not consistent on where we run it, its possible to miss some hours.  using the utc versions of these calls would fix it

Reply:
Could we use the file where we store with `fs` to store a timestamp that we read from the previous rebate to calculate this time window more accurately without relying on human actions?

Maybe overkill: I feel this script could use a db (noSql) or even google sheet controlled programatically to store this kind of information.

Reply:
again, I dont think any of these changes are worth it. as long as we run it from the same computer/ gcp instance each time it does not matter.

also remember that the voting rounds are longer than a day so its unlikely to miss anything.

what I will look into is forcing it to use a fixed timestamp within the script so that there is no margin for error irrespective of where you run the script from (i.e the script always uses UTC, for example).